src/components/visualizer.tsx

```diff
@@ -26,43 +31,64 @@ export default function Visualizer({ inVisualizer, current, goBack }: Visualizer
   // TODO: actual functions
   function start() {
     console.log('test start');
+    
+    let running = true;
+    let delayMilliseconds = simulationSpeed;
+    setRunState("Starting...");
+    const task = async function() {
+        await window.platform._temp_launchProgram("path");
+        setRunState("Started");
+        // TODO: improve cancellation
+        // possibly refactor this into some sort of controller class
+        while (running) {
+            const msg = await window.platform._temp_doStep();
+            console.log("received visualizer state", msg);
+            if (msg.type === 'process_end') {
+              setRunState("Finished");
+              return;
+            }
+            setConsoleOutput((console_out) => {
+              let new_out = console_out;
+              // Not actually lines
+              for (const printed_line of msg.printed_lines) {
+                new_out += printed_line;
+              }
+              return new_out;
+            });
+            
+            setVisualizerState(msg);
+            await delay(delayMilliseconds);
+        }
+    }();
+    visualizerController.current = {
+      stop: () => {running = false;},
+      setSpeed: (delay: number) => {delayMilliseconds = delay;}
+    };
```

Comment:
At some unknown point in the future, lets try and refactor this to a TS Class, if possible

Reply:
BTW this is only a comment for the future, not a change request, when you see this mark the conversation as resolved.

Reply:
I went ahead and made it a class. It's in src/components/visualizer/visualizerController.ts. It should probably be moved out of the components folder, but I wasn't sure where to put it.

Reply:
I think where it is works